components/brave_vpn/brave_vpn_connection_manager_mac.mm

```diff
@@ -0,0 +1,249 @@
+    }
+    LOG(ERROR) << "Error: storing password";
+  }
+
+  return status;
+}
+
+NEVPNProtocolIKEv2* CreateProtocolConfig(const BraveVPNConnectionInfo& info) {
+  NSString* hostname = [NSString stringWithUTF8String:info.hostname().c_str()];
+  NSString* username = [NSString stringWithUTF8String:info.username().c_str()];
+
+  NEVPNProtocolIKEv2* protocol_config = [[NEVPNProtocolIKEv2 alloc] init];
+  protocol_config.serverAddress = hostname;
+  protocol_config.serverCertificateCommonName = hostname;
+  protocol_config.remoteIdentifier = hostname;
+  protocol_config.enablePFS = YES;
+  protocol_config.disableMOBIKE = NO;
+  protocol_config.disconnectOnSleep = NO;
+  protocol_config.authenticationMethod =
+      NEVPNIKEAuthenticationMethodCertificate;  // to validate the server-side
+                                                // cert issued by LetsEncrypt
+  protocol_config.certificateType = NEVPNIKEv2CertificateTypeECDSA256;
+  protocol_config.useExtendedAuthentication = YES;
+  protocol_config.username = username;
+  protocol_config.passwordReference = GetPasswordRefForAccount(kBraveVPNKey);
+  protocol_config.deadPeerDetectionRate =
+      NEVPNIKEv2DeadPeerDetectionRateLow; /* increase DPD tolerance from default
+                                             10min to 30min */
+  protocol_config.useConfigurationAttributeInternalIPSubnet = false;
+
+  [[protocol_config IKESecurityAssociationParameters]
+      setEncryptionAlgorithm:NEVPNIKEv2EncryptionAlgorithmAES256];
+  [[protocol_config IKESecurityAssociationParameters]
+      setIntegrityAlgorithm:NEVPNIKEv2IntegrityAlgorithmSHA384];
+  [[protocol_config IKESecurityAssociationParameters]
+      setDiffieHellmanGroup:NEVPNIKEv2DiffieHellmanGroup20];
+  [[protocol_config IKESecurityAssociationParameters]
+      setLifetimeMinutes:1440];  // 24 hours
+  [[protocol_config childSecurityAssociationParameters]
+      setEncryptionAlgorithm:NEVPNIKEv2EncryptionAlgorithmAES256GCM];
+  [[protocol_config childSecurityAssociationParameters]
+      setDiffieHellmanGroup:NEVPNIKEv2DiffieHellmanGroup20];
+  [[protocol_config childSecurityAssociationParameters]
+      setLifetimeMinutes:480];  // 8 hours
+
+  return protocol_config;
+}
+
+}  // namespace
+
+// static
+BraveVPNConnectionManager* BraveVPNConnectionManager::GetInstance() {
+  static base::NoDestructor<BraveVPNConnectionManagerMac> s_manager;
+  return s_manager.get();
+}
+
+BraveVPNConnectionManagerMac::BraveVPNConnectionManagerMac() = default;
+BraveVPNConnectionManagerMac::~BraveVPNConnectionManagerMac() = default;
+
+void BraveVPNConnectionManagerMac::CreateVPNConnection(
+    const BraveVPNConnectionInfo& info) {
+  info_ = info;
+
+  if (StorePassword([NSString stringWithUTF8String:info_.password().c_str()],
+                    kBraveVPNKey) != errSecSuccess)
+    return;
+
+  NEVPNManager* vpn_manager = [NEVPNManager sharedManager];
+  [vpn_manager loadFromPreferencesWithCompletionHandler:^(NSError* error) {
```

Comment:
Block will strongly capture `vpn_manager`.

Reply:
Should I use `[NEVPNManager sharedManager]` also in completion handler instead of using `vpn_manager`?

Reply:
I double checked. You can ignore this one. The manager is static anyway so it doesn't matter if there is an extra reference to it. Usually we do `__weak` and `__strong`. Your solution would work too, but I would just ignore it as it isn't an issue since `sharedManager` returns static instance.